fix: align MCP tool ids for permissions

nujabse opened the pull request:
### What this PR does

Before this PR:
- MCP tools were listed with `mcp_` IDs but the agent SDK used `mcp__` tool names, so allowed tool toggles did not auto-approve.

After this PR:
- MCP tools are listed with `mcp__` IDs and legacy allowlist entries are normalized for agents and sessions.

Fixes # (none)

### Why we need it and why it was done in this way

The following tradeoffs were made:
- Normalize allowlists on read to keep existing data working without a migration.

The following alternatives were considered:
- Data migration to rewrite stored IDs, which would be heavier and more intrusive.

Links to places where the discussion took place:

### Breaking changes

None.

### Special notes for your reviewer

None.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: Write code that humans can understand and Keep it simple
- [x] Refactor: You have left the code cleaner than you found it (Boy Scout Rule)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A user-guide update was considered and is not required

### Release note

```release-note
Fix MCP tool auto-approval to respect allowed tool toggles by aligning tool IDs.
```

DeJeune commented:
🎯 Overall Assessment: APPROVE WITH REQUIRED CHANGES (Score: 6.5/10)

  ✅ What Works Well:

  - Correctly fixes the MCP tool ID mismatch bug (mcp_ vs mcp__)
  - Excellent backward compatibility approach - no database migration needed
  - Clean implementation with proper abstraction
  - No security vulnerabilities

  ❌ Critical Issues to Fix Before Merge:

  1. Missing Session List Normalization (BLOCKING)
    - SessionService.listSessions() doesn't call normalizeAllowedTools()
    - Creates inconsistency compared to getSession()
    - File: src/main/services/agents/services/SessionService.ts:204
  2. Zero Test Coverage (BLOCKING)
    - No unit tests for the normalization logic
    - No integration tests for the SDK tool approval flow
    - Need tests in src/main/services/agents/__tests__/BaseService.test.ts
  3. Missing Documentation (BLOCKING)
    - No inline comments explaining the ID format change
    - No migration guide for developers
    - Need to document the legacy vs. new format

  📝 Specific Code Changes Needed:

  Fix 1: Add normalization to SessionService.listSessions() (line ~204):
  ```ts
  const sessions = result.map((row) => this.deserializeJsonFields(row)) as GetAgentSessionResponse[]
  // ADD THIS:
  for (const session of sessions) {
    session.tools = await this.listMcpTools(session.agent_type, session.mcps)
    session.allowed_tools = this.normalizeAllowedTools(session.allowed_tools, session.tools)
  }

  return { sessions, total }
```
  Fix 2: Add comprehensive unit tests covering normalization, deduplication, and edge cases

  Fix 3: Add JSDoc documentation explaining the legacy vs. current format

  📊 Detailed Metrics:

  - Correctness: 8/10
  - Test Coverage: 2/10 ⚠️
  - Security: 10/10 ✅
  - Backward Compatibility: 10/10 ✅

nujabse commented:
Hi @DeJeune, thanks for the thorough review. I've addressed the blocking items:

  - Session list normalization: listSessions now calls listMcpTools + normalizeAllowedTools, matching getSession.
  - Tests: Added unit coverage for normalization/dedup/edge cases in src/main/services/agents/tests/BaseService.test.ts.
  - Docs: Added JSDoc in BaseService.normalizeAllowedTools explaining legacy  vs current format.

kangfenmao commented:
Great analysis and fix for the MCP tool permissions issue!

I have a suggestion for improvement: would it be better to unify the MCP tool ID generation by using `buildFunctionCallToolName()` from `@main/utils/mcp`? There is another PR #12186 that refactor this function to use the same `mcp__` prefix format.

Currently, the ID generation happens in two places:
1. `BaseService.listMcpTools()` (this PR) - uses template string `mcp__${id}__${tool.name}`
2. `buildFunctionCallToolName()` in `mcp.ts` (PR #12186) - uses `mcp__${server}__${tool}`

This could lead to inconsistencies. Would you consider depending on PR #12186 and using `buildFunctionCallToolName()` here for consistency?

Something like:
```typescript
import { buildFunctionCallToolName } from '@main/utils/mcp'

tools.push({
  id: buildFunctionCallToolName(server.name, tool.name),
  // ...
})
```

vaayne commented on src/main/utils/mcp.ts:
```diff
@@ -1,56 +1,28 @@
+ * @param toolName - The tool name from the server
+ * @returns A valid JS identifier in format mcp__{server}__{tool}, max 63 chars
+ */
+export function buildFunctionCallToolName(serverName: string, toolName: string): string {
```
Hi @DeJeune, I believe this is from my PR. @nujabse did not rebase, so it's here. 

I also think we can ignore the issue of two servers with the same name, as this is unlikely to happen. Even if it does occur, I don't think we should fix it in the code, as it adds more complexity than value.
